### PR TITLE
 update sync_diff_inspector: sync_diff_inspector v2.0

### PR DESCRIPTION
**Pull request by Liuxiaozhen12**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [x] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/7353 & https://github.com/pingcap/docs-cn/pull/7510
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 1]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
-# If it is set to false, data is checked.
-ignore-data-check = false
+```bash
+./bin/sync_diff_inspector --config=./config.toml
+```
 
-# If it is set to true, the table struct comparison is ignored.
-# If set to false, the table struct is compared.
-ignore-struct-check = false
+This command outputs a check report `summary.txt` in the `output-dir` of `config.toml` and the log `sync_diff.log`. In the `output-dir`, a folder named by the hash value of the `config. toml` file are also generated. This folder includes the checkpoint node information of breakpoint and the SQL file generated when the data is inconsistent.
```
Comment (hfxsd) [inline sync-diff-inspector/sync-diff-inspector-overview.md:163]:
```suggestion
This command outputs a check report `summary.txt` in the `output-dir` of `config.toml` and the log `sync_diff.log`. In the `output-dir`, a folder named by the hash value of the `config. toml` file is also generated. This folder includes the checkpoint node information of breakpoints and the SQL file generated when the data is inconsistent.
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 2]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
+        'output/fix-on-tidb2/'
+You can view the comparision details through 'output/sync_diff.log'
+```
 
-Run the following command:
+### Output file
 
-``` bash
-./bin/sync_diff_inspector --config=./config.toml
+The output file directory structure is as follows:
```
Comment (hfxsd) [inline sync-diff-inspector/sync-diff-inspector-overview.md:204]:
```suggestion
The directory structure of the output file is as follows:
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 3]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
 
-#### Log
+The running sync-diff-inspector periodically (every 10 seconds) prints the progress in checkpoint, which is located at `${output}/checkpoint/sync_diff_checkpoints.pb`, among which `${output}` is the value of `output-dir` in the `config.toml` file.
 
-The running sync-diff-inspector periodically (every 10 seconds) prints the progress in log in the following format:
+### Result
 
-```log
-[2020/11/12 17:47:00.170 +08:00] [INFO] [checkpoint.go:276] ["summary info"] [instance_id=target] [schema=test] [table=test_table] ["chunk num"=1000] ["success num"=80] ["failed num"=1] ["ignore num"=0]
+After the check is finished, sync-diff-inspector outputs a report, which is located at `${output}/summary.txt`, among which `${output}` is the value of `output-dir` in the `config.toml` file.
```
Comment (hfxsd) [inline sync-diff-inspector/sync-diff-inspector-overview.md:232]:
```suggestion
After the check is finished, sync-diff-inspector outputs a report. It is located at `${output}/summary.txt`, and `${output}` is the value of `output-dir` in the `config.toml` file.
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 4]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
-+ If the data is consistent, a log example is as follows:
+### SQL statements to fix inconsistent data
 
-    ```log
-    [2020/11/12 17:47:00.174 +08:00] [INFO] [report.go:80] ["check result summary"] ["check passed num"=1] ["check failed num"=0]
-    [2020/11/12 17:47:00.174 +08:00] [INFO] [report.go:87] ["table check result"] [schema=test] [table=test_table] ["struct equal"=true] ["data equal"=true]
-    [2020/11/12 17:47:00.174 +08:00] [INFO] [main.go:75] ["check data finished"] [cost=353.462744ms]
-    [2020/11/12 17:47:00.174 +08:00] [INFO] [main.go:69] ["check pass!!!"]
-    ```
+If different rows exist during the data checking process, the SQL statements to fix them will generate. If the data inconsistency exists in a chunk, a SQL file named by `chunk.Index` will generate. The SQL file is located at `${output}/fix-on-${instance}`, among which `${instance}` is the value of `task.target-instance` in the `config.toml` file.
```
Comment (hfxsd) [inline sync-diff-inspector/sync-diff-inspector-overview.md:253]:
```suggestion
If different rows exist during the data checking process, the SQL statements will be generated to fix them. If the data inconsistency exists in a chunk, a SQL file named by `chunk.Index` will be generated. The SQL file is located at `${output}/fix-on-${instance}`, and `${instance}` is the value of `task.target-instance` in the `config.toml` file.
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 5]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
-    ```log
-    [2020/11/12 18:16:17.068 +08:00] [INFO] [checkpoint.go:276] ["summary info"] [instance_id=target] [schema=test] [table=test1] ["chunk num"=1] ["success num"=0] ["failed num"=1] ["ignore num"=0]
-    [2020/11/12 18:16:17.071 +08:00] [INFO] [report.go:80] ["check result summary"] ["check passed num"=0] ["check failed num"=1]
-    [2020/11/12 18:16:17.071 +08:00] [INFO] [report.go:87] ["table check result"] [schema=test] [table=test_table] ["struct equal"=true] ["data equal"=false]
-    [2020/11/12 18:16:17.071 +08:00] [INFO] [main.go:75] ["check data finished"] [cost=319.849706ms]
-    [2020/11/12 18:16:17.071 +08:00] [WARN] [main.go:66] ["check failed!!!"]
-    ```
+- If the rows in the downstream database are missing, REPLACE statements will be applied
+- If the rows in the downstream database are redundant, DELETE statements will be applied
+- If some data of the rows in the downstream database is inconsistent, REPLACE statements will be applied and different columns will be marked with annotation in the SQL file
```
Comment (hfxsd) [inline sync-diff-inspector/sync-diff-inspector-overview.md:259]:
```suggestion
- If some data of the rows in the downstream database is inconsistent, REPLACE statements will be applied and inconsistent columns will be marked with annotation in the SQL file
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 6]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
-- Pay special attention to `table-rules`. If you configure `schema-pattern="test1"` and `target-schema="test2"`, the `test1` schema in the source database and the `test2` schema in the target database are compared. If the source database has a `test2` schema, this `test2` schema is also compared with the `test2` schema in the target database.
-- The generated `fix.sql` is only used as a reference for repairing data, and you need to confirm it before executing these SQL statements to repair data.
+- Pay special attention to `table-rules`. If you configure `schema-pattern="test1"`, `table-pattern = "t_1"`, `target-schema="test2"` and `target-table = "t_2"`, the `test1`.`t_1` schema in the source database and the `test2`.`t_2` schema in the target database are compared. Sharding is enabled by default in sync-diff-inspector, so if the source database has a `test2`.`t_2` table,  the `test1`.`t_1` table and `test2`.`t_2` table in the source database serving as sharding are compared with the `test2`.`t_2` table in the target database.
```
Comment (hfxsd) [inline sync-diff-inspector/sync-diff-inspector-overview.md:280]: 出现这种问题会对用户有什么影响呢？如果有风险，可以加个说明明确指出。

**Review by hfxsd — APPROVED** [review]
LGTM

File: sync-diff-inspector/route-diff.md [hunk 7]
```diff
@@ -6,58 +6,58 @@ aliases: ['/docs/dev/sync-diff-inspector/route-diff/','/docs/dev/reference/tools
...
+[routes.rule1]
+schema-pattern = "test_1"      # # Matches the schema name of the data source. Supports the wildcards "*" and "?"
+table-pattern = "*"            # Matches the table name of the data source. Supports the wildcards "*" and "?"
+target-schema = "test_2"       # The name of the schema in the target database
+target-table = "t_2"           # The name of the target table
 ```
+
+## Note
+
+If `t_2` exists in the upstream database, the downstream databse also compares this table.
```
[the remark this answers is not in the document]
Comment (Liuxiaozhen12) [reply sync-diff-inspector/route-diff.md:63]: 我理解的也是这个意思。@leavrth PTAL

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 8]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
-* In TiDB and MySQL, `FLOAT`, `DOUBLE` and other floating-point types are implemented differently, so checksum might be calculated differently. If the data checks are inconsistent due to these data types, set `ignore-columns` to skip checking these columns.
+* In TiDB and MySQL, `FLOAT`, `DOUBLE` and other floating-point types are implemented differently. `FLOAT` and `DOUBLE` respectively take 6 and 15 significant digits for calculating checksum. If you do not want to use this feature, set `ignore-columns` to skip checking these columns.
 
 * Support checking tables that do not contain the primary key or the unique index. However, if data is inconsistent, the generated SQL statements might not be able to repair the data correctly.
 
-### Database privilege
+## Database privileges for sync-diff-inspector
 
-sync-diff-inspector needs to obtain the information of table schema, to query data, and to build the `checkpoint` database to save breakpoint information. The required database privileges are as follows:
+sync-diff-inspector needs to obtain the information of table schema to query data. The required database privileges are as follows:
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/sync-diff-inspector-overview.md:43]:
```suggestion
sync-diff-inspector needs to obtain the information of table schema and to query data. The required database privileges are as follows:
```

File: sync-diff-inspector/sync-diff-inspector-overview.md (hunk 6 above)
Comment (Liuxiaozhen12) [reply sync-diff-inspector/sync-diff-inspector-overview.md:280]: @Leavrth PTAL~

File: sync-diff-inspector/route-diff.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Leavrth) [reply sync-diff-inspector/route-diff.md:63]: 这里少了，应该是如果上游数据库存在`test_2`.`t_2`，也会检查

File: sync-diff-inspector/sync-diff-inspector-overview.md (hunk 6 above)
Comment (Leavrth) [reply sync-diff-inspector/sync-diff-inspector-overview.md:280]: 一般来说，如果没有rules，那么上游数据库会匹配下游数据库相同名称的表，相当于每个表都有个默认的自己匹配自己的rule。但是加上rules并不会去掉这个默认的匹配规则，也就是如上面说的，如果用户想要做 test1.t_1 -> test2.t_2 的匹配，但是默认还存在 test2.t_2 -> test2.t_2 的匹配，那么上游存不存在 test2.t_2 的结果是不一样的。

File: sync-diff-inspector/route-diff.md [hunk 9]
```diff
@@ -6,58 +6,58 @@ aliases: ['/docs/dev/sync-diff-inspector/route-diff/','/docs/dev/reference/tools
...
+[routes.rule1]
+schema-pattern = "test_1"      # Matches the schema name of the data source. Supports the wildcards "*" and "?"
+table-pattern = "*"            # Matches the table name of the data source. Supports the wildcards "*" and "?"
+target-schema = "test_2"       # The name of the schema in the target database
+target-table = "t_2"           # The name of the target table
 ```
+
+## Note
+
+If `t_2` exists in the upstream database, the downstream databse also compares this table.
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/route-diff.md:63]:
```suggestion
If `test_2`.`t_2` exists in the upstream database, the downstream databse also compares this table.
```
```suggestion
If `t_2` exists in the upstream database, the downstream databse also compares this table.
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 10]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
 
-The configuration of sync-diff-inspector consists of three parts:
+The configuration of sync-diff-inspector consists of the following parts:
 
-- `Global config`: General configurations, including log level, chunk size, number of threads to check.
-- `Tables config`: Configures the tables for checking. If some tables have a certain mapping relationship between the upstream and downstream databases or have some special requirements, you must configure these tables.
+- `Global config`: General configurations, including number of threads to check, whether to export SQL statement to fix inconsistent tables, whether to campare the data and so on.
 - `Databases config`: Configures the instances of the upstream and downstream databases.
+- `Tables config`: Special configurations for specific tables, including specified ranges, columns to be ignored and so on (optional).
+- `Routes`: Rules for upstream multiple schema names to regularly match downstream single schema names (optional).
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/sync-diff-inspector-overview.md:61]:
```suggestion
- `Routes`: Rules for upstream multiple schema names to match downstream single schema names (optional).
```
```suggestion
- `Routes`: Rules for upstream multiple schema names to regularly match downstream single schema names (optional).
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 11]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
+- `Task config`: Configures the tables for checking. If some tables have a certain mapping relationship between the upstream and downstream databases or have some special requirements, you must configure these tables.
 
 Below is the description of a complete configuration file:
 
 ``` toml
 # Diff Configuration.
 
 ######################### Global config #########################
+# The number of goroutines created to check data. The number of connections between upstream and downstream databases are slightly greater than this value.
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/sync-diff-inspector-overview.md:70]:
```suggestion
# The number of goroutines created to check data. The number of connections between sync-diff-inspector and upstream/downstream databases is slightly greater than this value.
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 12]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
-# Uses "chunk-size" to set the size of a chunk.
-chunk-size = 1000
+# Only compares the table structure instead of the data.
+check-struct-only = false
 
-# The number of goroutines created to check data
-check-thread-count = 4
+######################### Datasource config #########################
+[data-sources]
+[data-sources.mysql1] # mysql1 is the only ID for the database instance. It is used in the following `task.source-instances/task.target-instance` files.
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/sync-diff-inspector-overview.md:81]:
```suggestion
[data-sources.mysql1] # mysql1 is the only ID for the database instance. It is used for following `task.source-instances/task.target-instance` configuration.
```

File: sync-diff-inspector/sync-diff-inspector-overview.md [hunk 13]
```diff
@@ -26,244 +26,256 @@ This guide introduces the key features of sync-diff-inspector and describes how
...
+    host = "127.0.0.1"
+    port = 4000
+    user = "root"
+    password = ""
+    # Uses the snapshot feature. If enabled, historical data is used for comparison
+    # snapshot = "386902609362944000"
+
+######################### Table config #########################
+# Special configurations for specific tables. The tables to be configured must be in `task.target-check-tables`.
+[table-configs.config1] # config1  is the only ID for this configuration. It is used in the following `task.target-configs`.
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/sync-diff-inspector-overview.md:97]:
```suggestion
[table-configs.config1] # config1  is the only ID for this configuration. It is used for the following `task.target-configs` configuration.
```

File: sync-diff-inspector/shard-diff.md [hunk 14]
```diff
@@ -21,196 +19,119 @@ Below is a complete example of the sync-diff-inspector configuration.
...
+
+    target-instance = ["tidb0"]
+
+    # The tables of downstream databases to be compared. Each table needs to contain the schema name and the table name, separated by '.'
+    target-check-tables = ["test.table-0"]
 ```
+
+## Note
+
+If `table_0` exists in the upstream database, the downstream databse also compares this table.
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/shard-diff.md:137]:
```suggestion
If `test.table_0` exists in the upstream database, the downstream databse also compares this table.
```

File: sync-diff-inspector/route-diff.md [hunk 15]
```diff
@@ -6,58 +6,58 @@ aliases: ['/docs/dev/sync-diff-inspector/route-diff/','/docs/dev/reference/tools
...
+[routes.rule1]
+schema-pattern = "test_1"      # Matches the schema name of the data source. Supports the wildcards "*" and "?"
+table-pattern = "*"            # Matches the table name of the data source. Supports the wildcards "*" and "?"
+target-schema = "test_2"       # The name of the schema in the target database
+target-table = "t_2"           # The name of the target table
 ```
+
+## Note
+
+If `test_2`.`t_2` exists in the upstream database, the downstream databse also compares this table.
```
Comment (Liuxiaozhen12) [inline sync-diff-inspector/route-diff.md:63]:
```suggestion
If `test_2`.`t_2` exists in the upstream database, the downstream database also compares this table.
```

**Review by Leavrth — APPROVED** [review]

**Pull request merged**